README.md

```diff
@@ -215,12 +215,80 @@ Body: {
 
 ### Job Applications
 
+#### Fetch all Job Applications for a user
+Request:
+```
+GET /api/v1/users/:user_id/job_applications
+```
+Headers:
+```
+{
+  "Authorization": "Bearer <your_token_here>"
+}
```

Comment:
Wouldn’t it be redundant to add the user id the user id is already being decoded from the token. It can already be inferred that the user is provided using the key.

Reply:
The comment above is not related to the code snippet you highlighted, but to answer your question on the path was added:
```
GET   /api/v1/users/:user_id/job_applications(.:format) 
```
was chosen because it aligns with several key considerations:
1. **Scalability**: While the `user_id` can be inferred from the token for the authenticated user, explicitly including it in the route allows for future scalability. For example, this structure supports scenarios where administrators or other users might need to interact with job applications belonging to other users.

2. **Explicitness**: Including the `user_id` in the route provides clarity and makes the relationship between the user and their job applications explicit. This is particularly beneficial for developers working on the app, as it eliminates any ambiguity and makes the API structure easier to follow.

3. **RESTful Convention**: In RESTful design, nested routes are commonly used to represent relationships between resources. Since `job_applications` belong to a specific `user`, the nested route reflects this relationship accurately.

While it might seem redundant when the user_id can be decoded from the token, this explicit approach enhances readability and makes the API more intuitive, especially in cases where the app grows to support additional user roles or functionalities.

Reply:
While these are all good points, the only way anybody will need to use the user id is being an admin or some type of editor while everything else will stay confidential between any user, because it wouldn’t make for a user to edit another user. If we are talking about scenarios. Adding the user id could also put risk on that users information. If we are using tokens and authentication to the routes a restful api using routes like api/v1/companies is more common practice because the server already decodes the user out out of the token, not an endpoint that involves the user directly in the endpoint.

Reply:
Let me clarify some points. There are no security risk at play here. The user_id is never publicly exposed. We currently are utilizing authentication with tokens, and using flat routes like /api/v1/job_applications does not provide additional security since the user context must still be validated.

Reply:
Thanks for the clarification. I understand the points you've raised about scalability, explicitness, and RESTful conventions, and I agree they do have merit, especially in cases where additional user roles or future functionalities might need to be considered.

Efficiency and Redundancy: As you've pointed out, the user ID can already be decoded from the token. Including it in the endpoint doesn't add any additional value or clarity since the server can already infer the user context. We would just be repeating the same information. This also follows the principle of not exposing redundant information in the URL, which can sometimes be a good practice to keep endpoints clean and focused on the resource itself.

Cleaner API Design: Keeping the endpoint simpler and using /api/v1/job_applications instead of /api/v1/users/:user_id/job_applications aligns more with how token-based authentication is typically used in RESTful APIs. By decoding the token and inferring the user, we make the API cleaner and easier to maintain, especially as we scale. This also makes the system easier to test, as there is no reliance on the user ID being passed explicitly.

Maintainability: As our app grows, it's better to handle user context internally (via the token) rather than relying on an explicit user ID in the route. This way, if we need to support other use cases or change how we handle users in the future, it won't require a refactor of the entire API structure.

Reply:
I’m with Melchor on this issue. John, you’re right that there’s no inherent security risk in including user_id in the route as long as the user context is validated server-side. With proper token-based authentication, both flat and nested routes can be secure.

That said, I still see value in using flat routes like /api/v1/job_applications for simplicity and to avoid redundancy. Since the token already provides user context, including user_id in the route feels unnecessary unless we’re explicitly supporting scenarios like admin or editor access to other users’ data. However, I would argue against scenarios where an admin or another user could access someone else’s private data. Job applications, contacts, and similar resources should remain private to the user. Admins could have control over user accounts as a whole (e.g., the ability to delete a user) but not the user’s private data.

For now, I believe flat routes better align with the app’s current needs and keep the API easier to use and maintain. If we anticipate adding roles or permissions that require accessing another user’s resources in the future, the nested structure could become relevant, but I think our current approach is the right fit.

Reply:
I respectfully disagree with the notion that including user_id in the route introduces redundancy. There’s a misconception here—redundancy implies unnecessary duplication, but the user_id in the route serves a different purpose than the Authorization token. While the token authenticates the requester, the user_id explicitly defines whose data is being accessed, making the relationship between users and their job applications clear. This explicitness aligns with RESTful principles and ensures the API is readable, intuitive, and scalable for future scenarios, such as roles like admins or editors. Flat routes rely on implicit assumptions, which can limit flexibility and clarity as the app evolves. Including user_id provides explicit context without any actual redundancy.

Reply:
This has been a great conversation, and I’m really glad we had the chance to discuss this. While I still feel the way I do about the approach, I respect the team’s decision and am glad we’re unified moving forward. It’s great to see us working together to make the best decisions for the project.